editor/scene/2d/particles_2d_editor_plugin.cpp

```diff
@@ -61,154 +63,258 @@ void GPUParticles2DEditorPlugin::_add_menu_options(PopupMenu *p_menu) {
+				r_valid_colors.write[i * 4 + 0] = mask_img_ptr[(point.y * mask_img_size.width + point.x) * 4 + 0];
+				r_valid_colors.write[i * 4 + 1] = mask_img_ptr[(point.y * mask_img_size.width + point.x) * 4 + 1];
+				r_valid_colors.write[i * 4 + 2] = mask_img_ptr[(point.y * mask_img_size.width + point.x) * 4 + 2];
+				r_valid_colors.write[i * 4 + 3] = mask_img_ptr[(point.y * mask_img_size.width + point.x) * 4 + 3];
+			}
+		}
+	}
+
+	if (direction_mode == DIRECTION_MODE_TEXTURE) {
+		Ref<Image> normal_img;
+		normal_img.instantiate();
+		err = ImageLoader::load_image(direction_img_path_line_edit->get_text(), normal_img);
+		ERR_FAIL_COND_MSG(err != OK, vformat("Error loading image '%s'.", direction_img_path_line_edit->get_text()));
+
+		if (normal_img->is_compressed()) {
+			normal_img->decompress();
+		}
+		normal_img->convert(Image::FORMAT_RGB8);
+		ERR_FAIL_COND(normal_img->get_format() != Image::FORMAT_RGB8);
+		Size2i normal_img_size = normal_img->get_size();
+		ERR_FAIL_COND(normal_img_size.width == 0 || normal_img_size.height == 0);
+		ERR_FAIL_COND_MSG(normal_img_size != mask_img_size, "Mask and Normal texture must have the same size.");
+
+		Vector<uint8_t> normal_img_data = normal_img->get_data();
+		const uint8_t *normal_img_ptr = normal_img_data.ptr();
+
+		for (int i = 0; i < valid_point_count; ++i) {
+			const Point2i point = r_valid_positions.get(i);
+			const uint8_t normal_r = normal_img_ptr[(point.y * normal_img_size.width + point.x) * 3 + 0];
+			const uint8_t normal_g = normal_img_ptr[(point.y * normal_img_size.width + point.x) * 3 + 1];
+
+			Vector2 normal;
+			normal.x = static_cast<float>(normal_r) / 255.0f - 0.5f;
+			normal.y = static_cast<float>(normal_g) / 255.0f - 0.5f;
+
+			normal.normalize();
+
+			r_valid_normals.write[i] = normal;
+		}
 	}
 
-	r_valid_positions.resize(vpc);
+	r_valid_positions.resize(valid_point_count);
 	if (!r_valid_normals.is_empty()) {
-		r_valid_normals.resize(vpc);
+		r_valid_normals.resize(valid_point_count);
 	}
 }
 
 Particles2DEditorPlugin::Particles2DEditorPlugin() {
-	file = memnew(EditorFileDialog);
+	file_dialog = memnew(EditorFileDialog);
 
 	List<String> ext;
 	ImageLoader::get_recognized_extensions(&ext);
 	for (const String &E : ext) {
-		file->add_filter("*." + E, E.to_upper());
+		file_dialog->add_filter("*." + E, E.to_upper());
 	}
 
-	file->set_file_mode(EditorFileDialog::FILE_MODE_OPEN_FILE);
-	EditorNode::get_singleton()->get_gui_base()->add_child(file);
-	file->connect("file_selected", callable_mp(this, &Particles2DEditorPlugin::_file_selected));
+	file_dialog->set_file_mode(EditorFileDialog::FILE_MODE_OPEN_FILE);
+	file_dialog->connect("file_selected", callable_mp(this, &Particles2DEditorPlugin::_file_selected));
 
-	emission_mask = memnew(ConfirmationDialog);
-	emission_mask->set_title(TTR("Load Emission Mask"));
+	emission_mask_dialog = memnew(ConfirmationDialog);
+	emission_mask_dialog->set_title(TTR("Load Emission Mask"));
```

Comment:
```suggestion
	emission_mask_dialog->set_title(TTRC("Load Emission Mask"));
```
Same in everything in the constructor.

Reply:
See the discussion in https://github.com/godotengine/godot/issues/104574
We don't have it documented yet AFAIK.